README.md
GoHA
====

Extremely light and powerful Node JS reverse proxy and forward-proxy


# Philosophy

  - Follows KISS (**Keep it simple, stupid**) principle, lighter than every other competitors
  - Focus on **High Performance** and **High availability**
  - Easy to fork and modify

# Features

  - Flexible and easy routing
  - Easy to deploy and upgrade without downtime
  - Secured by default in a carefully systemd sandbox
  - Seamless SSL Support (HTTPS -> HTTP proxy)
  - Automatic HTTP to HTTPS redirects
  - Works with Let's Encrypt, activate https as soon as a certificate is available
  - Multi-thread (cluster by default)
  - DNS Failover
  - Websockets, HTTP2
  - Zero Point of Failure (multi-server master-slave config)
  - Zero downtime: register and unregister routes, add certificates without restarting
  - Passes tests of heavily used `node-http-proxy` module but without memory leaks and with better performance
  - Reload-safe: if config file is broken, it keeps previous config in memory and in the disk (zero downtime)
  - (TODO) Includes awesome statistics
    - top-10 slowest queries
    - histogram per hour, per response time
    - send real-time statistics to Netdata



# Getting-started

## In production with systemd (Ubuntu/Debian ONLY)

GoHA relies on systemd to run with a high level of security and availability.

It provides automatic deployment scripts and a CLI for administration

```bash
  # Download the binary
  wget https://github.com/Ideolys/goha/releases/download/1.0.0-beta.2/goha-linux-1-0-0-beta-2
  # To install it, execute the installation command directly from the binary and follow instructions
  ./goha-linux-1-0-0-beta-2 install
```

Now GoHA is installed and running. See the Configuration part below to configure the proxy.

> By default, the service is installed in `/usr/local/bin/goha` and run with `goha` user.
> The working directory, where the configuration is stored, is `/var/www/goha`.
> It is possible to overwrite this values with environment variables `GOHA_USER` and `GOHA_WORKDIR`.

## For development or testing purpose

- For testing purpose

```bash
  # If you want to just execute "in place" for testing purpose
  # Create a working directory
  mkdir ~/goha
  # Download executable
  wget https://github.com/Ideolys/goha/releases/download/1.0.0-beta.2/goha-linux-1-0-0-beta-2
  # And start goha directly in the directory (=the working directory)
  ./goha-linux-1-0-0-beta-2 go
```

- For development:

```bash
  # Clone the source code and run:
  npm install
  # start the proxy locally without systemd
  bin/goha go
  # tests
  npm test
  # build binary
  npm run build
```


# Command Line Interface

```bash
  goha [commands]

  #  Production commands. For Linux only, GoHA must be installed before:
  #
  #    start          : Start the proxy
  #    stop           : Stops the proxy
  #    reload         : Reloads the configuration or upgrade seamlessly (no socket lost, no packet lost)
  #    log            : Shows logs in realtime
  #    restart        : Restarts completely the proxy with service interruption
  #    --version [-v] : Get current version
  #    --help [-h]    : Show this help
  #
  #  Installation and tests commands:
  #
  #    install  : Install the proxy as a systemd service (Linux only)
  #               Options:
  #                 --non-interactive : install without user interaction
  #                 --no-start        : do not start or restart service
  #    go       : Start the proxy without systemd, only for test purpose
  #
```




# Configuration 

GoHA can be configured with three methods
  - **[method 1]** : a static JSON file
  - **[method 2]** : an HTTP API, which updates the static JSON file. **Available only if the method 1 is used.**
  - **[method 3]** : a dynamic Javascript file which build and exports the JSON configuration file.

Configuration files are stored in the working directory in `$GOHA_WORKDIR` (`/var/www/goha` by default).

Here is the file structure of the working directory:

```bash
  |- config.json          # [method 1] user-defined configuration file
  |- config.js            # [method 3] user-defined javascript which exports the configuration
  |- .config-runtime.json # last valid configuration file currently used in production (DO NOT MODIFY).
  |- backup
    |- config-20210505121001.json # automatic backup of previous configuration file
    |- config-20210603121011.json # GoHA keeps all history for the moment
  | middleware
    |- loadBalancing.js  # (TODO) load balancing middleware functions    
    |- onRequest.js      # (TODO) on request middleware
    |- onResponse.js     # (TODO) on response middleware
  | public 
    |- 404.html          # (TO_IMPROVE) default mainteance or 404 error page

```


## [method 1] config.json

Here are all options available in `config.json` 

```js
  {
    "port" : 80,
    "portSSL" : 443,    // GoHA reads Let's Encript certificate automatically in /etc/letsencrypt/live
    "portAdmin" : 3000, // REST API to update configuration remotely (method 2) and dashboard + monitoring URLs
    "domains" : {
      // every request coming to blabla.company.net will be routed to http://100.100.100.100:8101
      "blabla.company.net" : "http://100.100.100.100:8101",
      "toto.company.net"   : "http://100.100.100.101:8102",
      // Underscore char can be used to deactivate a redirection wihout removing the line from the JSON [TODO]
      "_titi.company.net"   : "http://100.100.100.101:8102",
      // Add custom headers like x-schema for ideos multitenant
      "saas-client-1.company.net" : { 
        "backends" : [{ "url" : "http://100.100.100.101:8102" }],
        "headers"  : {"x-schema" : "5000" }
      },
      // Manage Load balancing
      "saas-client-2.company.net" : {
        "backends" : [
          { "url" : "http://100.100.10.101:8102"                    }, // backend server 1
          { "url" : "http://100.100.10.101:8103"                    }, // backend server 2, ...
          { "url" : "http://100.100.10.100:8103", "isReady" : false }, // if isReady is false, no requests will be sent to this backend
          { "url" : "http://100.100.10.100:8104", "version" : "2"   }  // routing by version
        ],
        "versioning" : {
          "header"  : "App-Version", // case-insensitive version header name
          "default" : "0",           // default version if client has no version header
        },
        // Default load balancing method used if loadBalanceFn is undefined : Least Connection.
        // But you can overwrite it with this optional function.
        // For each request, a loop calls this function for each backend where isReady is true
        // It sends the request to the backend for which the function returns the lowest number
        // Be carefull, this function is stringified so you cannot use variable coming from outside
        "loadBalancingFn" : "", // [TODO]
        // custom error page if saas-client-2.company.net is not available
        "errorPage" : "custom404.html", // [TODO]
        // accept traffic coming from this interface only
        "listen" : "192.161.1.1" // [TODO]
      }
    },
    // default error page if a website is not available
    "errorPage" : "404.html",
    // Automatically generate new domains with duplication rules.
    // Ex. duplicate all domains URL replacing ".company.net" by ".1.alt.company.com" and ".2.alt.company.com"
    "alternativeDomains" : {
      ".company.net" : [".1.alt.company.com", ".2.alt.company.com"]
    }
  }
```

## [method 2] REST API + Dashboard administration

GoHA listens `portAdmin` for administration if `portAdmin` is defined.

**⚠️ Be careful, this port should not be publicly exposed even if its protected ⚠️** . You should allow access only through a VPN.

*List of APIs:*

- `GET  /      ` : [HTML] show a simple dashboard (TODO) 
- `GET  /config` : [JSON] get current runtime config
- `POST /config` : [JSON] overwrite existing configuration. All missing element are deleted ⚠️
- `PUT  /config` : [JSON] merge with existing configuration (update and add elements only)
- `GET  /status` : [HTML] status of GoHA (TODO)
- `GET  /metric` : [HTML] open metrics (TODO)


### PUT /config

The configuration sent by API is merged with the existing one. It adds missing domains, modify existing attributes and it does not delete anything

**Parameter**

The JSON format is exactly the same as the configuration file on the disk.

**Body**

```js
  {
    "domains" : {
      "blabla.company.net" : "http://100.100.100.100:8101", // add a domain
    }
  }
```

**Response**:

It returns the whole config in "data" attribute

```js
  {
    "data": {
      "domains" : {
        "blabla.company.net" : "http://100.100.100.100:8101",
        "toto.company.net"   : "http://100.100.100.101:8102"  // existing config
      },
    "message" : "success message"
  }
```


## [method 3] Javascript mode

The config file can be a dynamic Javascript file. So you can write code to generate a dynamic config file.

This file must export the config object.

```javascript
  let config = {}; // same config as explained earlier
  module.exports = config;
```

> ⚠️ `config.json` is ignored if there is a `config.js` in the working directly.




## Internal workflow

The master does little things. Everything is done in workers.

- The master starts
  - It copies `config.json` to `.config-runtime.json` if the latter does not exists
  - It starts workers

- When a worker starts
  - It read only `.config-runtime.json`

- When the signal to reload the configuration is received
  - The master catches the signal
  - The master sends a message to one worker to test the new `config.json` 
  - If the worker can read the configuration with success, it sends a "success" signal to the master
  - If the master receives the "success" signal, it replaces `.config-runtime.json` by the content of `config.json` and creates a backup of the previous config in `backup` directory
  - The master sends a signal to all workers to really reload the configuration by reading `.config-runtime.json`


`.config-runtime.json`  can be different from `config.json` if the latter contains an error and cannot be loaded.
In that case, `.config-runtime.json` is used as a backup to run the last working configuration in production (Very useful when the machine restarts).



# Logging

You can use `goha log` to logs in live mode. It is a shortcut of `journalctl -n 500 -f -u $GOHA_SERVICE_NAME`


Log format : `"[TYPE]"  "LEVEL"  "message"`

- `[TYPE]` can be 
  - `[APP]` : application message, such as config reload, start/stop info
  - `[REQ]` : request logs

- `LEVEL` can be
  - `INFO`  : general info messages
  - `ERROR` : error messages
  - `WARN`  : carning messages

- `message` can be
  - for `[APP]` message, general info
  - for `[REQ]` see format below


> GoHA relies on systemd to lotates logs


Logs output (TODO)
```
  [REQ] INFO 192.168.1.1 www.mysite.com http://192.161.1.10:3001 GET /css/main.css unique_request_id
  [REQ] INFO 192.168.1.1 www.mysite.com http://192.161.1.10:3001 GET /css/main.css unique_request_id 
```




# Version routing

GoHA can route packets according to version header. Each backend must have a version.

```js
  'saas-client-2.company.net' : {
    backends : [
      { url : 'http://100.100.10.101:8102', version : '1'          },
      { url : 'http://100.100.10.100:8104', version : '2'          }, // matches with version '2', '2.0', '2.1.1-beta'
      { url : 'http://100.100.10.100:8105', version : '2.1'        }, // matches with version '2.1', '2.1.2', '2.1.2-beta'
      { url : 'http://100.100.10.100:8106', version : '2010-01-01' }  // matches with version '2010-01-01', '2010-01-01.1020'
    ],
    versioning : {
      header  : 'App-Version', // case-insensitive version header name
      default : '1'            // default version if client has no version header
    }
  }
```

If there is a conflicts, GoHA selects the backend which has the longest version string.
In the example above, if the client has a header `"App-Version" : "2.1"`, the backend which has the port 8105 is choosen.




# How to activate SSL?


GoHA redirects automatically the client to https if there is a corresponding certificate for the domain in `/etc/letsencrypt/live`.

GoHA does not generate and renew cerificates himself (For the moment, it must be done manually or in a cron tab).

Let's encrypt does not accept wildcard certificates, so we must generate one certificate per sub-domain/domain.

Let's encrypt cannot generate more than 5 certificates per week (rate limit), so take your time ;).


1. Install Let's encrypt

    On ubuntu 16.04 LTS (source : https://certbot.eff.org/#ubuntuxenial-other)
    
    ```
    sudo apt-get install software-properties-common
    sudo add-apt-repository ppa:certbot/certbot
    sudo apt-get update
    sudo apt-get install certbot
    ```

2. Generate a certificate for one domain

    ```
      certbot certonly --webroot -w /var/www/goha/public -d failover.myapp.net
    ```
    
    **When the first certificate is created, change owner of these directory to make it GoHA-accessible:**
    
    ```
      sudo chown -R ubuntu:ubuntu /etc/letsencrypt/
      sudo chown -R ubuntu:ubuntu /var/log/letsencrypt/
      sudo chown -R ubuntu:ubuntu /var/lib/letsencrypt/
    ```

3. Reload GoHA

    `goha reload`



### How to renew SSL Certificates ?

Each month, renew all certificates in a cron tab with

```
certbot renew
cd /var/www/goha/
goha reload
```

### DNS Let encrypt

https://buzut.net/certbot-challenge-dns-ovh-wildcard/






# Load balancing

GoHA can be used for load balancing between many backends servers defined in an array like this:

```js
'saas-client-1.company.net' : {
  backends : [
    { url: 'http://100.100.100.10:8102'                    },
    { url: 'http://100.100.100.101:8102' , isReady : false },// if isReady is false, no requests will be sent to this backend
    { url: 'http://100.100.100.101:8103'                   } // by default : isReady : true
  ],
  loadBalancingFn : (req, backend) => { return backend.nbConnection }
}
```

GoHA uses "the least connections" algorithm by default.
Read [this doc](https://d2c.io/post/haproxy-load-balancer-part-2-backend-section-algorithms) about all possible load balancing algorithms

But you can overwrite with your own function or one of these built-in functions :

```js
  /**
   * If the request is a GET, it redirects the URL `/render/<base64IPv4` to the
   * backend which matches with the <base64IPv4> encoded with Buffer.from('127.0.0.2'.padEnd(15, ' ')).toString('base64');
   */
  loadBalancing.selectBackendUsingBase64IP(/^\/render\/(.*)/)
```


# Notes

#### Doc of postintall command line

```bash
# If the file cert.pem does not exists
(    [ ! -f \"./cert.pem\" ] 
# And npm install is executed without the option "--production" (used by Gihtub Actions to speed up npm install, avoiding devDeps)
  && [ -z \"$npm_config_production\" ]
) 
# Generate a certificate for development
&& openssl req -new -newkey rsa:4096 -days 3650 -nodes -x509 -subj \"/C=FR/ST=FR/L=Paris/O=Goha/OU=Goha/CN=localhost/emailAddress=no@goha.io\" -keyout ./privkey.pem -out ./cert.pem 
# Else, print a message
|| echo \"Default certificate already generated or production install\" "
```




CHANGELOG.md
# GoHA

### 1.0.0-beta.0
  - First version of GoHA


test/test.helper.js
const helper  = require('../lib/helper.js');
const params  = require('../lib/params.js');

describe('helper', function () {

  describe('sanitize', function () {
    it('should sanitize input', () => {
      helper.eq(helper.sanitize('   CONNECTION   '), 'connection');
      helper.eq(helper.sanitize('keeP-alive '), 'keep-alive');
      helper.eq(helper.sanitize(), '');
      helper.eq(helper.sanitize(null), '');
      helper.eq(helper.sanitize(22), '');
    });
  });

  describe('uid', function () {
    it('should return a unique id', function (done) {
      var _uids   = [];
      var _nbUids = 1000;
      for (var i = 0; i < _nbUids; i++) {
        _uids.push(helper.uid());
      }
      for (var j = _nbUids - 1; j >= 0; j--) {
        var _uid   = _uids[j];
        var _found = false;

        for (var k = 0; k < j; k++) {
          if (_uid === _uids[k]) {
            _found = true;
            break;
          }
        }
        helper.eq(_found, false);
        _uids.splice(j, 1);
        if (!_uids.length) {
          done();
        }
      }
    });
  });

  describe('createHttpHeader', function () {
    it('should generate HTTP headers and insert the prefix and return a buffer', () => {
      let _headers = {
        'user-agent' : 'curl/7.22.0',
        host         : '127.0.0.1:8000',
        accept       : '*/*'
      };
      let _res = helper.createHttpHeader('prefix', _headers);
      helper.eq(Buffer.isBuffer(_res), true);
      helper.eq(_res.toString('utf8'), 'prefix\r\nuser-agent: curl/7.22.0\r\nhost: 127.0.0.1:8000\r\naccept: */*\r\n\r\n' );
    });
    it('should accept array in headers', () => {
      let _headers = {
        host    : '127.0.0.1:8000',
        cookies : ['bla', 'other']
      };
      let _res = helper.createHttpHeader('HTTP', _headers);
      helper.eq(_res.toString('utf8'), 'HTTP\r\nhost: 127.0.0.1:8000\r\ncookies: bla\r\ncookies: other\r\n\r\n' );
    });
    it('should never crash if headers is  are string is null or undefined', () => {
      helper.eq(helper.createHttpHeader('HTTP', {}).toString('utf8'), 'HTTP\r\n\r\n' );
      helper.eq(helper.createHttpHeader('HTTP', []).toString('utf8'), 'HTTP\r\n\r\n' );
      helper.eq(helper.createHttpHeader('HTTP', null).toString('utf8'), 'HTTP\r\n\r\n' );
      helper.eq(helper.createHttpHeader('HTTP', 2).toString('utf8'), 'HTTP\r\n\r\n' );
      helper.eq(helper.createHttpHeader('HTTP').toString('utf8'), 'HTTP\r\n\r\n' );
    });
  });

  describe('getRequestHeaders', function () {
    it('should copy headers (except host, why ????????????????) and set the correct x-forwarded-* headers', function () {
      let _request = {
        connection : {
          remoteAddress : '192.168.1.2',
          remotePort    : '8080'
        },
        headers : {
          host   : '192.168.1.2:8080',
          accept : '*/*'
        }
      };
      _request[params.UID] = '112asdazdaz';
      let _clone = helper.cloneHeaderForProxy(_request);
      helper.eq(_clone, {
        accept              : '*/*',
        'x-forwarded-for'   : '192.168.1.2',
        'x-forwarded-proto' : 'http',
        'x-forwarded-host'  : '192.168.1.2:8080',
        'x-request-id'      : '112asdazdaz'
      });
    });
    it('should overwrite existing x-forwarded and add a unique global id with x-request-id', function () {
      let _request = {
        connection : {
          remoteAddress : '192.168.1.2',
          remotePort    : '8080'
        },
        headers : {
          host                : '192.168.1.2:8080',
          'x-forwarded-for'   : '10.168.1.3',
          'x-forwarded-proto' : 'https',
          'x-forwarded-host'  : '10.168.1.3'
        }
      };
      _request[params.UID] = '3312asdazdaz';
      let _clone = helper.cloneHeaderForProxy(_request);
      helper.eq(_clone, {
        'x-forwarded-for'   : '192.168.1.2',
        'x-forwarded-proto' : 'http',
        'x-forwarded-host'  : '192.168.1.2:8080',
        'x-request-id'      : '3312asdazdaz'
      });
    });
  });
});


lib/worker.js
const http        = require('http');
const https       = require('https');
const cluster     = require('cluster');
const proxy       = require('./proxy');
const requestIp   = require('request-ip');
const url         = require('url');
const fs          = require('fs');
const path        = require('path');
const helper      = require('./helper');
const params      = require('./params');
const config       = require('./config');
const certificates = require('./certificates');
const { pipeline, finished } = require('stream');
const log         = helper.logger('APP');
const stat        = helper.logger('REQ');
let   serverAdmin = null;
let   serverHttp  = null;
let   serverHttps = null;
// certificates per domain
let defaultCertificates   = {};
let certificatesPerDomain = {};
let targetsPerDomain      = {};
let currentConfig         = {};

// without a global agent, connection is closed after each request, reducing performance (x 2)
// https://github.com/nodejitsu/node-http-proxy/issues/929
var keepAliveAgent = new http.Agent({ keepAlive : true });
// create the proxy
// var proxy = httpProxy.createProxyServer({
//   agent : keepAliveAgent
// });


/**
 * Process messages coming form master
 *
 * @param  {Object} msg
 */
function onPrimaryMessage (msg) {
  switch (msg) {
    case params.DRY_RELOAD_ONE_WORKER:
      readConfig(false) && process.send(params.RELOAD_ALL_WORKERS);
      break;
    case params.RELOAD_ONE_WORKER:
      readConfig(true);
      break;
    case params.SHUTDOWN_ONE_WORKER:
      log.info('Shutdown workers. Waiting remaining connections...');
      // TODO better manager if or use stopListening
      serverHttp.close(() => {
        serverHttps.close(() => {
          log.info('Exit');
          process.exit(0);
        });
      });
      break;
    default:
      break;
  }
}

/**
 * Starts a listening.
 */
function startListening () {
  // init SSL options for https.createServer function
  var sslOptions = {
    SNICallback : function (hostname, callback) {
      callback(null, certificatesPerDomain[hostname]);
    },
    // add default certificate when client does not provide SNI (mandatory)
    key  : defaultCertificates.privkey,
    cert : defaultCertificates.cert
  };

  if (currentConfig.portAdmin) {
    serverAdmin = http.createServer(handleAdminRequest);
    serverAdmin.listen(currentConfig.portAdmin);
    log.info(`Listen administration port ${currentConfig.portAdmin}`);
  }
  serverHttp  = http.createServer(handleRequest);
  serverHttps = https.createServer(sslOptions, handleRequest);
  serverHttp.listen(currentConfig.port);
  serverHttps.listen(currentConfig.portSSL);
  log.info(`Listen port ${currentConfig.port}`);
  log.info(`Listen port SSL ${currentConfig.portSSL}`);

  serverHttp.on('upgrade' , onUpgradeWS);
  serverHttps.on('upgrade', onUpgradeWS);
}

/**
 * Stops a listening.
 */
function stopListening () {
  if (serverHttp) {
    serverHttp.close();
  }
  if (serverHttps) {
    serverHttps.close();
  }
}

/**
 * Handle incoming request
 *
 * @param  {Object}  req     The request
 * @param  {Object}  res     The resource
 */
function handleRequest (req, res) {
  // set unique id for this request
  req[params.UID]      = helper.uid();
  var _hostWithoutPort = helper.getHostWithoutPort(req);
  var _target          = loadBalance(targetsPerDomain, _hostWithoutPort, req, res);
  var _ip              = requestIp.getClientIp(req);
  stat.info( _ip + '\t' + req.headers.host + '\t' + (_target !== undefined ? _target.url : 'unknown') + ' : ' + req.method + ' ' + req.url + '\t' + req[params.UID]);

  // if this an http request, without SSL
  if (req.connection.encrypted === undefined) {
    var _hasCertificate = certificatesPerDomain[_hostWithoutPort];
    // if this is a lets encrypt validation request
    if (/^\/.well-known\/acme-challenge/.test(req.url) === true) {
      return sendFilePublic(req, res);
    }
    // if there is a certificate for this domain, redirect the client to https
    if ( _hasCertificate !== undefined ) {
      // let _url = `https://${_hostWithoutPort}${(currentConfig.portSSL === 443 ? '' : ':' + currentConfig.portSSL)}/req.url);
      var _url = 'https://' + path.join(_hostWithoutPort + (currentConfig.portSSL === 443 ? '' : ':' + currentConfig.portSSL) , req.url);
      stat.info('redirect to ' + _url);
      res.writeHead(302, { Location : _url });
      return res.end();
    }
  }
  // if target is known, proxy it
  if (_target !== undefined) {
    const { port, hostname } = new URL(_target.url); // TODO REMOVE new URL
    // TODO manage htpt https, repalce slow new URL
    return proxy.web(req, res, {protocol : 'http', hostname : hostname, port : port, headers : _target.headers }, (err, rq, rs, hd, obj) => {
      if ( !err ) {
        return;
      }
      if (err.statusCode !== 503 &&  err.statusCode !== 502 || obj !== undefined /* a websocket */) { // TODO confirm
        return res.end();
      }
      // If it fails, try another backend in xx miliseconds
      var _retryTimeout = 50;
      var _otherTarget = loadBalance(targetsPerDomain, _hostWithoutPort, req, res, _target.url);
      stat.info('request-retry' + _ip + '\t' + req.headers.host + '\t' + _otherTarget.url + ' : ' + req.method + ' ' + req.url);
      // If the backend is the same, increase timeout
      if (_otherTarget.url === _target.url) {
        _retryTimeout = 10000;
      }
      setTimeout(() => {
        const { port, hostname } = new URL(_otherTarget.url); // TODO REMOVE new URL
        proxy.web(req, res, { protocol : 'http', hostname : hostname, port : port, headers : _target.headers }, (err) => {
          // TODO test
          if ( !err ) {
            return;
          }
          return res.end();
        });
      }, _retryTimeout);
    });
  }
  // otherwise return error
  stat.error('Unknown target for ' + _hostWithoutPort);
  res.writeHead(404, {'Content-Type' : 'text/plain'});
  res.end('Page not found');
}

/**
 * Listener for upgrade event
 * @param {Object} req
 * @param {Object} socket
 * @param {Object} head
 */
function onUpgradeWS (req, socket, head) {
  var _hostWithoutPort = helper.getHostWithoutPort(req);

  var _target = targetsPerDomain[_hostWithoutPort];
  if (_target === undefined || !(_target.backends instanceof Array) ) {
    return req.end();
  }
  _target = _target.backends[0];

  var _ip = requestIp.getClientIp(req);
  stat.info('[request-upgrade] ' + _ip + '\t' + req.headers.host + '\t' + _target.url + ' : ' + req.url);
  // return proxy.ws(req, socket, head, { target : _target.url, xfwd : true });
  const { port, hostname } = new URL(_target.url); // TODO REMOVE new URL
  return proxy.ws(req, socket, head, { protocol : 'ws', hostname : hostname, port : port, headers : _target.headers });
}

/**
 * If there are multiple backends, select the backend which have the least connection
 * to send the request
 *
 * @param  {Object} targetPerDomainConfig object from config
 * @param  {String} hostWithoutPort       hostname
 * @param  {Object} req
 * @param  {Object} res
 * @param  {Object} otherThan             avoid returning this backend URL is possible
 * @return {Object}                       target selected
 */
function loadBalance (targetPerDomainConfig, hostWithoutPort, req, res, otherThan) {
  var _target = targetPerDomainConfig[hostWithoutPort];
  if (_target === undefined || !(_target.backends instanceof Array) ) {
    return;
  }
  if (_target.backends.length === 1) {
    return _target.backends[0];
  }
  let _isVersionning = typeof(_target.versioning) === 'object' && typeof(_target.versioning.header) === 'string';
  let _wantedVersion = '';
  if (_isVersionning === true) {
    _wantedVersion = (req.headers[_target.versioning.header] !== undefined) ? req.headers[_target.versioning.header] : _target.versioning.default;
  }

  // If the target is an array of backend
  let _lowestValue = Number.MAX_SAFE_INTEGER;
  let _lowestConnectionBackend = 0;
  // select the backend which have the lowest value returned by loadBalancingFn (by default lowest number of connections)
  for (let i = 0; i < _target.backends.length; i++) {
    let _candidateBackend = _target.backends[i];
    if ( typeof(_target.loadBalancingFn ) === 'string' ) {
      _target.loadBalancingFn = new Function('req', 'backend', 'return (' + _target.loadBalancingFn + ')(req, backend);');
    }
    let _value = _target.loadBalancingFn(req, _candidateBackend);
    if (_value === Number.MIN_SAFE_INTEGER) {
      _lowestConnectionBackend = i;
      break;
    }
    if (_isVersionning === false || _candidateBackend.version === _wantedVersion) {
      if (_value < _lowestValue && _candidateBackend.isReady === true && _candidateBackend.url !== otherThan) {
        _lowestValue = _value;
        _lowestConnectionBackend = i;
      }
    }
  }
  // TODO manage if all server are not ready!
  let _chosenTarget = _target.backends[_lowestConnectionBackend];
  _chosenTarget.nbConnection++;

  res.on('finish', function () {
    _chosenTarget.nbConnection--;
    // protection when config is reloaded and if there are some closed connections later
    if (_chosenTarget.nbConnection < 0) {
      _chosenTarget.nbConnection = 0;
    }
  });

  return _chosenTarget;
}

// TODO c'est maintenant géré dans la callback du proxy.web et proxy.ws
// TODO gérer les headerSent dans le callback et vérifer que l'erreur est bien retournée
function handleError (err, req, res) {
  var _ip = requestIp.getClientIp(req);
  console.error('[request' + (!res.writeHead ? '-upgrade' : '') + '] ' +  _ip + ' -> ' + req.headers.host + ' :' + (!res.writeHead ? '' : ' ' + req.method) + ' ' + req.url);
  console.error(err);
  var _response = 'Internal server error';

  // WS res has no writeHead method
  if (!res.writeHead) {
    res.headersSent = true;
  }

  if (!res.headersSent) {
    // we must call writeHead only once. Do not send header if it has been already sent by the proxied app
    // if (req.xhr || req.headers.accept.indexOf('json') > -1) {
    //  res.writeHead(500, {'Content-Type': 'application/json'});
    //  _response = JSON.stringify({ success: false, error: 'Internal server error' });
    // } else {
    res.writeHead(500, {'Content-Type' : 'text/plain'});
    // }
  }
  return res.end(_response);
}

/**
 * Manage administration requests
 *
 * @param  {Object}  req  The request
 * @param  {Object}  res  The resource
 */
function handleAdminRequest (req, res) {
  const _method = req.method.toUpperCase();
  const _url    = url.parse(req.url, true);
  const _route  = _method + ' ' + _url.pathname;
  const _ip     = requestIp.getClientIp(req);
  log.info( _ip + '\t' + req.headers.host + '\t' + req.method + ' ' + req.url + '\t');
  switch (_route) {
    case 'GET /':
      helper.httpResponse(res, 200, 'GoHA Hello World\n');
      break;
    case 'GET /config':
      helper.httpResponse(res, 200, { data : currentConfig });
      break;
    case 'POST /config':
    case 'PUT /config':
      helper.getJSONData(req, (err, body) => {
        if (err) {
          return helper.httpResponse(res, 500, { message : err });
        }
        let _newConfig = body;
        if (_route.startsWith('PUT') === true ) {
          // merge mode
          _newConfig = config.merge(currentConfig, body);
        }
        if (!_newConfig) {
          return helper.httpResponse(res, 200, { message : 'Nothing to update' });
        }
        fs.writeFile(params.CONFIG_FILE, JSON.stringify(_newConfig, null, 2), (err) => {
          let _configStatus = readConfig(false, false); // try to reload config file
          if (err || _configStatus === false)  {
            return helper.httpResponse(res, 500, { message : 'Invalid configuration ' + err });
          }
          process.send(params.RELOAD_ALL_WORKERS);
          helper.httpResponse(res, 200, { data : _newConfig, message : 'The configuration will be updated in few seconds' });
        });
      });
      break;
    default:
      helper.httpResponse(res, 404, '404 Not Found\n');
  }
}

/**
 * When Lets Encrypt generate a certificate, it writes a file in ./public/.well-known/acme-challenge
 * We must return it, otherwise Lets Encrypt cannot validate the certificate
 */
// function sendFilePublic (req, res) {
//   let _uri = url.parse(req.url).pathname; // TODO replace
//   _uri = _uri.replace(/\.+/g, '.'); // security to avoid going up in hierarchy
//   let _filename = path.join(params.WORKING_DIR, 'public', _uri);
//
//   console.log('[request] LetsEncrypt trying to validate challenge: ', _filename);
//
//   fs.access(_filename, fs.constants.R_OK, function (err) {
//     if (err) {
//       console.error(err);
//       res.writeHead(404, {'Content-Type' : 'text/plain'});
//       res.write('404 Not Found\n');
//       return res.end();
//     }
//     res.writeHead(200);
//     fs.createReadStream(_filename, 'binary').pipe(res);
//   });
// }

function sendFilePublic (req, res) {
  let _uri = url.parse(req.url).pathname; // TODO replace
  _uri = _uri.replace(/\.+/g, '.'); // security to avoid going up in hierarchy
  let _filename = path.join(params.WORKING_DIR, 'public', _uri);

  stat.info(`Sending file ${_filename}\t${req[params.UID]}`);

  fs.access(_filename, fs.constants.R_OK, function (err) {
    if (err) {
      stat.error(`File ${_filename} not found\t${req[params.UID]}`);
      res.writeHead(404, {'Content-Type' : 'text/plain'});
      res.write('404 Not Found\n');
      return res.end();
    }
    res.writeHead(200);
    const _file = fs.createReadStream(_filename, 'binary');
    pipeline(_file, res, (err) => {
      if (err) {
        stat.error(`File ${_filename} not sent correctly\t${req[params.UID]}`);
      }
    });
  });
}


/**
 * Reads a configuration.
 *
 * @param      {boolean}  [isReadingRuntimeConfig=false]  Indicates if reading runtime configuration.
 *                        If false, it tries to read the new config file (not runtime one) without applying it
 * @return     {boolean}  true  if the configuration is valid
 *                        false if the configuration contains error
 */
function readConfig (isReadingRuntimeConfig = false, isUpdatingCertificate = true) {
  let _newCertificatesPerDomain = {};
  let _newDefaultCertificates = {};
  if (isUpdatingCertificate === true) {
    try {
      _newCertificatesPerDomain = certificates.getLetsEncryptCertificates();
      _newDefaultCertificates = certificates.getDefaultCertificate();
    }
    catch (e) {
      log.error('An error occurs when reading let\'s encrypt certificates or default certificates. ', e);
      log.error('Keep using in-memory let\'s encrypt certificates certificates');
      log.warn('Reload CANCELED');
      return false;
    }
  }

  let _newConfigParsed = config.getConfig(currentConfig, isReadingRuntimeConfig);
  if (!_newConfigParsed) {
    log.error('Config file is broken or does not contain any domain to listen');
    log.error('Fix config.js/json and try again');
    log.warn('Reload CANCELED');
    return false;
  }

  if (isReadingRuntimeConfig === true) {
    // apply config only if we read the runtime file
    currentConfig         = _newConfigParsed.configSource;
    targetsPerDomain      = _newConfigParsed.targetsPerDomain;
    log.info('Config reloaded with SUCCESS!');
    if (isUpdatingCertificate === true) {
      certificatesPerDomain = _newCertificatesPerDomain;
      defaultCertificates   =  _newDefaultCertificates;
      certificates.resolveWildcardCertificate(certificatesPerDomain, targetsPerDomain);
      log.info('Certificates reloaded with SUCCESS!');
    }
  }
  return true;
}


// Start automatically if it is worker (does not start in tests)
if (cluster.isWorker === true) {
  readConfig(true);
  startListening();
}
// listen messages coming from Primary process
process.on('message', onPrimaryMessage);


module.exports = {
  startListening,
  stopListening,
  handleRequest,
  onPrimaryMessage,
  // only for test purpose
  set targetsPerDomain (o)      { targetsPerDomain      = o; }, // eslint-disable-line
  set currentConfig (o)         { currentConfig         = o; }, // eslint-disable-line
  set certificatesPerDomain (o) { certificatesPerDomain = o; }, // eslint-disable-line
  set defaultCertificates (o)   { defaultCertificates   = o; }  // eslint-disable-line
};


